src/libraries/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.AddSub.cs

```diff
@@ -9,58 +9,131 @@ namespace System.Numerics
+                }
             }
 
-            bits[left.Length] = (uint)carry;
         }
 
         public static void Add(ReadOnlySpan<uint> left, ReadOnlySpan<uint> right, Span<uint> bits)
         {
+            Debug.Assert(right.Length >= 1);
             Debug.Assert(left.Length >= right.Length);
             Debug.Assert(bits.Length == left.Length + 1);
 
-            int i = 0;
-            long carry = 0L;
-
             // Switching to managed references helps eliminating
             // index bounds check...
-            ref uint leftPtr = ref MemoryMarshal.GetReference(left);
             ref uint resultPtr = ref MemoryMarshal.GetReference(bits);
+            ref uint rightPtr = ref MemoryMarshal.GetReference(right);
+            ref uint leftPtr = ref MemoryMarshal.GetReference(left);
+
+            nint i = 0;
+            nint upperBound = right.Length;
+            long carry = 0;
 
             // Executes the "grammar-school" algorithm for computing z = a + b.
             // While calculating z_i = a_i + b_i we take care of overflow:
             // Since a_i + b_i + c <= 2(2^32 - 1) + 1 = 2^33 - 1, our carry c
             // has always the value 1 or 0; hence, we're safe here.
 
-            for ( ; i < right.Length; i++)
+            do
             {
-                long digit = (Unsafe.Add(ref leftPtr, i) + carry) + right[i];
-                Unsafe.Add(ref resultPtr, i) = unchecked((uint)digit);
-                carry = digit >> 32;
+                carry += Unsafe.Add(ref leftPtr, i);
+                carry += Unsafe.Add(ref rightPtr, i);
+                Unsafe.Add(ref resultPtr, i) = unchecked((uint)carry);
+                carry >>= 32;
+                i++;
+            } while (i < upperBound);
+            upperBound = left.Length;
+
+            if (upperBound <= CopyToThreshold)
+            {
+                for ( ; i < upperBound; i++)
+                {
+                    carry += Unsafe.Add(ref leftPtr, i);
+                    Unsafe.Add(ref resultPtr, i) = unchecked((uint)carry);
+                    carry >>= 32;
+                }
+
+                Unsafe.Add(ref resultPtr, upperBound) = unchecked((uint)carry);
             }
-            for ( ; i < left.Length; i++)
+            else
             {
-                long digit = left[i] + carry;
-                Unsafe.Add(ref resultPtr, i) = unchecked((uint)digit);
-                carry = digit >> 32;
+                for ( ; i < upperBound; )
+                {
+                    carry += Unsafe.Add(ref leftPtr, i);
+                    Unsafe.Add(ref resultPtr, i) = unchecked((uint)carry);
+                    i++;
+                    carry >>= 32;
+                    if (carry == 0) break;
+                }
+
+                Unsafe.Add(ref resultPtr, upperBound) = unchecked((uint)carry);
+
+                if (i < upperBound)
+                {
+                    CopyTail(left, bits, unchecked((int)i));
+                }
```

Comment:
this code seems identical with part of `Add(ReadOnlySpan<uint> left, uint right, Span<uint> bits)`. If possible we should refactor it to helper method and enforce inlining, so the code is reused but there is no perf penalty.